search/utils.py

```diff
@@ -0,0 +1,366 @@
+    text_keys = (
+        'title', 'titlesort', 'resource_type', 'description', 'content_xml',
+        'content_stripped',
+    )
+    for key in text_keys:
+        try:
+            rec[key] = rec[key].decode('utf-8')
+        except AttributeError:
+            pass  # Python 3
+
+    return rec
+
+
+def clear_index():
+    """Wipe the index."""
+    if CONN.indices.exists(INDEX_NAME):
+        CONN.indices.delete(INDEX_NAME)
+        CONN.indices.create(INDEX_NAME)
+        CONN.indices.refresh()
+    create_mapping()
+
+    # re-index all existing LearningResource instances:
+    index_resources(LearningResource.objects.all())
+
+
+class SearchResults(object):
+    """
+    Helper class for elasticsearch_dsl search results.
+
+    This is because the actual search result has nested types, which
+    would be cumbersome to access from arbitrary calling functions.
+    """
+    def __init__(self, search):
+        """Get raw search result from Elasticsearch."""
+        self._search = search
+
+    def count(self):
+        """Total records matching the query."""
+        return self._search.count()
+
+    def page_count(self):
+        """Total number of result pages."""
+        count = self._search.count() / PAGE_LENGTH
+        if self._search.count() % PAGE_LENGTH > 0:
+            count += 1
+        return int(count)
+
+    def get_page(self, page):
+        """Return paginated results."""
+        start = (page - 1) * PAGE_LENGTH
+        end = start + PAGE_LENGTH
+        return self._search[start:end].execute().hits
+
+    def all(self):
+        """Return all results in a generator."""
+        for i in range(self.page_count()):
+            for hit in self.get_page(i+1):
+                yield hit
+
+    def __getitem__(self, i):
+        """Return result by index."""
+        return self._search[i:i+1].execute().hits[0]
+
+
+def create_mapping():
+    """
+    Create the Elasticsearch mapping.
+
+    Notes on the "index" values:
+        analyzed: normal, will be split on whitespace for indexing
+            (partial matches can be made during queries)
+        not_analyzed: no split; matches must be exact
+        no: not indexed at all, but will be returned in search results
+    """
+
+    # Create the index if it doesn't exist.
+    if not CONN.indices.exists(INDEX_NAME):
+        CONN.indices.create(INDEX_NAME)
+    # Delete the mapping if an older version exists.
+    if CONN.indices.exists_type(index=INDEX_NAME, doc_type=DOC_TYPE):
+        CONN.indices.delete_mapping(index=INDEX_NAME, doc_type=DOC_TYPE)
+
+    mapping = Mapping(DOC_TYPE)
+
+    mapping.field("course", "string", index="not_analyzed")
+    mapping.field("description_path", "string", index="no")
+    mapping.field("description", "string", index="analyzed")
+    mapping.field("preview_url", "string", index="no")
+    mapping.field("repository", "string", index="not_analyzed")
+    mapping.field("resource_type", "string", index="analyzed")
```

Comment:
If we're faceting on resource_type wouldn't we want an exact match here? Not sure if it matters, I don't think any resource types have spaces at all